homeassistant/components/light/__init__.py

```diff
@@ -999,58 +1010,73 @@ def _light_internal_convert_color(
     @property
     def state_attributes(self) -> dict[str, Any] | None:
         """Return state attributes."""
-        if not self.is_on:
-            return None
-
         data: dict[str, Any] = {}
         supported_features = self.supported_features
+        supported_color_modes = self._light_internal_supported_color_modes
         color_mode = self._light_internal_color_mode
 
-        if color_mode not in self._light_internal_supported_color_modes:
+        if color_mode not in supported_color_modes:
             # Increase severity to warning in 2021.6, reject in 2021.10
             _LOGGER.debug(
                 "%s: set to unsupported color_mode: %s, supported_color_modes: %s",
                 self.entity_id,
                 color_mode,
-                self._light_internal_supported_color_modes,
+                supported_color_modes,
             )
 
-        data[ATTR_COLOR_MODE] = color_mode
+        data[ATTR_COLOR_MODE] = color_mode if self.is_on else None
 
-        if color_mode in COLOR_MODES_BRIGHTNESS:
-            data[ATTR_BRIGHTNESS] = self.brightness
+        if brightness_supported(self.supported_color_modes):
+            data[ATTR_BRIGHTNESS] = (
+                self.brightness
+                if self.is_on and color_mode in COLOR_MODES_BRIGHTNESS
+                else None
+            )
```

Comment:
This is very chatty, but I don't really have a good idea for how to make it more terse.
Maybe some helper:
```py
    def __filtered_brightness(self) -> int | None:
        if not self.is_on or self._light_internal_supported_color_modes not in COLOR_MODES_BRIGHTNESS:
           return None
        return self.brightness

     ...

    if brightness_supported(self.supported_color_modes):
        data[ATTR_BRIGHTNESS] = self.__filtered_brightness()
```
Although that seems wildly inefficient because `_light_internal_supported_color_modes` would be called many times.

Or maybe something like this:
```py
    def __filtered_brightness(self, color_mode: str | None) -> int | None:
        if color_mode not in COLOR_MODES_BRIGHTNESS:
           return None
        return self.brightness

     ...

    if brightness_supported(self.supported_color_modes):
        data[ATTR_BRIGHTNESS] = self.__filtered_brightness(color_mode)
```


We also don't really like the ternary operator that much in core, this would be preferred:
```suggestion
        if brightness_supported(self.supported_color_modes):
            if color_mode in COLOR_MODES_BRIGHTNESS:
                data[ATTR_BRIGHTNESS] = self.brightness
            else:
                data[ATTR_BRIGHTNESS] = None
```